server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -155,115 +154,14 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
-                                requiredPipeline,
-                                defaultPipeline);
-                            throw new IllegalArgumentException(message);
-                        }
-                        final String pipeline;
-                        if (requiredPipeline != null) {
-                            pipeline = requiredPipeline;
-                        } else {
-                            pipeline = Objects.requireNonNullElse(defaultPipeline, IngestService.NOOP_PIPELINE_NAME);
-                        }
-                        indexRequest.setPipeline(pipeline);
-                    }
-
-                    if (requestPipeline != null) {
-                        if (requestCanOverridePipeline == false) {
-                            final String message = String.format(
-                                Locale.ROOT,
-                                "request pipeline [%s] can not override required pipeline [%s]",
-                                requestPipeline,
-                                requiredPipeline);
-                            throw new IllegalArgumentException(message);
-                        } else {
-                            indexRequest.setPipeline(requestPipeline);
-                        }
-                    }
-
-                    if (IngestService.NOOP_PIPELINE_NAME.equals(indexRequest.getPipeline()) == false) {
-                        hasIndexRequestsWithPipelines = true;
-                    }
-                    /*
-                     * We have to track whether or not the pipeline for this request has already been resolved. It can happen that the
-                     * pipeline for this request has already been derived yet we execute this loop again. That occurs if the bulk request
-                     * has been forwarded by a non-ingest coordinating node to an ingest node. In this case, the coordinating node will have
-                     * already resolved the pipeline for this request. It is important that we are able to distinguish this situation as we
-                     * can not double-resolve the pipeline because we will not be able to distinguish the case of the pipeline having been
-                     * set from a request pipeline parameter versus having been set by the resolution. We need to be able to distinguish
-                     * these cases as we need to reject the request if the pipeline was set by a required pipeline and there is a request
-                     * pipeline parameter too.
-                     */
-                    indexRequest.isPipelineResolved(true);
-                } else if (IngestService.NOOP_PIPELINE_NAME.equals(indexRequest.getPipeline()) == false) {
+                boolean indexRequestHasPipeline = resolveRequiredOrDefaultPipeline(actionRequest, indexRequest, metaData);
+                if (indexRequestHasPipeline) {
                     hasIndexRequestsWithPipelines = true;
```

Comment:
Maybe `hasIndexRequestWithPipelines |= indexRequestHasPipeline`. This will short-circuit if `hasIndexRequestWithPipelines` is already true and the branch predictor will love it.